lib/synapse/service_watcher/zookeeper_poll.rb

```diff
@@ -0,0 +1,81 @@
+require 'synapse/service_watcher/base'
+require 'synapse/service_watcher/zookeeper'
+
+require 'zk'
+require 'thread'
+require 'timeout'
+
+class Synapse::ServiceWatcher
+  class ZookeeperPollWatcher < ZookeeperWatcher
+    def start
+      log.info 'synapse: ZookeeperPollWatcher starting'
+
+      @thread = nil
+      @mutex = Mutex.new
+      @mutex.lock
```

Comment:
What do you want to do here? If you want a single thread to periodically polling, the `lock` should happen inside `until` loop?

Reply:
Actually, not sure whether you need a lock here or not. Could you simply use `@should_exit` to control it? ZookeeperPollWatcher < ZookeeperWatcher < BaseWatcher, so you should have `@should_exit` already. Given this, I think it might be better for you to extend directly from `BaseWatcher` directly. For common part, you could extract them to another class called `ZookeeperCommon`? see how other watchers, i.e., `DnsWatcher`, works.